R/DMMParam.R

```diff
@@ -0,0 +1,199 @@
+NULL
+
+#' @export
+#' @rdname DMMParam-class
+setClass("DMMParam", contains="BlusterParam", slots=c(k="integer_OR_NULL", 
+                                                      type="character",
+                                                      transposed="logical",
+                                                      seed = "integer_OR_NULL"))
+
+#' @export
+#' @rdname DMMParam-class
+DMMParam <- function(k=NULL, type=NULL, transposed=FALSE, seed=NULL) {
+    # Filling in missing values with the defaults.
+    current <- list(k=k, type=type, seed=seed)
+    notpresent <- vapply(current, is.null, FALSE)
+    if (any(notpresent)) {
+        defaults <- list(k=1:3, 
+                         type="laplace", 
+                         seed=runif(1, 0, .Machine$integer.max))
+        current[notpresent] <- defaults[notpresent]
+    }
+    
+    # Formatting data
+    if (!is.null(current$k)) {
+        current$k <- as.integer(current$k)
+    }
+    if (!is.null(current$seed)) {
+        current$seed <- as.integer(current$seed)
+    }
+
+    new("DMMParam", k=current$k, type=current$type, 
+        transposed=transposed, seed=current$seed)
+}
+
+#' @export
+setMethod("show", "DMMParam", function(object) {
+    callNextMethod()
+    if (length(object@k) > 1) {
+        cat(sprintf("k: %s\n", paste(range(object@k), collapse = ":")))
+    } else {
+        cat(sprintf("k: %s\n", object@k))
+    }
+    cat(sprintf("type: %s\n", object@type))
+    cat(sprintf("transposed: %s\n", object@transposed))
+    cat(sprintf("seed: %s\n", object@seed))
+})
+
+setValidity2("DMMParam", function(object) {
+    msg <- character(0)
+    if (!is.integer(object@k) ||
+        length(object@k) < 1 ||
+        any(object@k <= 0)) {
+        msg <- c(msg, "'k' must be a strictly positive integer vector.")
+    }
+    if (length(object@type) > 1) {
+        msg <- c(msg, "'type' must be a string")
+    } else if (!(object@type %in% c("laplace","AIC","BIC"))) {
+        msg <- c(msg, "'type' must be equal to 'laplace', 'AIC', or 'BIC'")
+    }
+    
+    if (length(msg) > 0) {
+        msg
+    } else {
+        TRUE
+    }
+})
+
+#' @export
+#' @rdname DMMParam-class
+#' @importFrom package function
+setMethod("clusterRows", c("ANY", "DMMParam"), function(x, 
+                                                        BLUSPARAM, 
+                                                        full=FALSE) {
+    k <- BLUSPARAM[["k"]]
+    type <- BLUSPARAM[["type"]]
+    transposed <- BLUSPARAM[["transposed"]]
+    seed <- BLUSPARAM[["seed"]]
+
+    if (!transposed) {
+        x <- t(x)
+    }
+    
+    dmm <- .get_dmm(x, k=k, seed = seed)
+    
+    # Finding optimal number of clusters if necessary
+    if (length(k) > 1) {
+        k <- bestDMMFit(dmm, type)
+    } 
+    
+    # Get the index corresponding to k in dmm list
+    i <- which(sapply(dmm, 
+                      function(x, k) ncol(DirichletMultinomial::mixture(x)) == k, 
+                      k=k))[1]
```

Comment:
There seems to be little bit extra maybe... does this work

i <- which(
        sapply(dmm, function(x) ncol(DirichletMultinomial::mixture(x)) == k))